### ISSUE TITLE
multiple_matching_tokens The cache contains multiple tokens satisfying the requirements. Call AcquireToken again providing more requirements such as authority or account.

### ISSUE THREAD

**Issue by r-delgadillo** [AzureAD/microsoft-authentication-library-for-js#5134]
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.12.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Public or Confidential Client?

Confidential

### Description

After adding 2 new MSAL flows for acquireTokenByRefreshToken and acquireTokenOnBehalfOf we started seeing some errors in existing workflows that had no issues (acquireTokenByClientCredential). The error we are receiving from MSAL is "multiple_matching_tokens". We only see this issue in acquireTokenByClientCredential for our singleton client.

We added `authority` to all of our client codes as suggested by the error message below but we are still receiving these errors. 

We noticed 2 similar incidents that were supposed to fix this issue:
[#4486](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4486)
[#4691](https://github.com/AzureAD/microsoft-authentication-library-for-js/pull/4691)

### Error Message

>"stack": "ClientAuthError: multiple_matching_tokens: The cache contains multiple tokens satisfying the requirements. Call AcquireToken again providing more requirements such as authority or account..\n    at ClientAuthError.AuthError [as constructor] (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:496:24)\n    at new ClientAuthError (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:790:28)\n    at Function.ClientAuthError.createMultipleMatchingTokensInCacheError (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:885:16)\n    at ClientCredentialClient.readAccessTokenFromCache (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:6036:35)\n    at ClientCredentialClient.<anonymous> (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:5996:50)\n    at step (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:79:23)\n    at Object.next (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:60:53)\n    at /app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:53:71\n    at new Promise (<anonymous>)\n    at __awaiter (/app/node_modules/@azure/msal-node/node_modules/@azure/msal-common/dist/index.cjs.js:49:12)",
>     "message": "**multiple_matching_tokens**: **The cache contains multiple tokens satisfying the requirements. Call AcquireToken again providing more requirements such as authority or account..**",
>     "errorCode": "multiple_matching_tokens",
>     "errorMessage": "The cache contains multiple tokens satisfying the requirements. Call AcquireToken again providing more requirements such as authority or account..",
>     "subError": "",
>     "name": "ClientAuthError"
>   }

### MSAL Configuration

```javascript
auth: {
                clientId: "9edfcdd9-0bc5-4bd4-b287-c3afc716aac7"
                authority: "https://login.microsoftonline.com/organizations",
                clientCertificate: {
                    thumbprint: pem.thumbprint,
                    privateKey: pem.privateKey,
                    x5c: pem.certificate,
                },
            },
```

### Relevant Code Snippets

```javascript
this._client.acquireTokenByClientCredential({
  "scopes": [
    "https://management.azure.com/.default"
  ],
  "authority": "https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47"
}),
```

### Reproduction Steps

N/A

### Expected Behavior

Expectation would be for `acquireTokenBYClientCredential` to return a valid access token matching the passed in requirements.

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

None (Server)

### Source

Internal (Microsoft)

**Comment by bgavrilMS** [comment]
Does this always repro? Or do you have to have a few tokens in the cache already?

Possible workaround: Can you try to specify the authority only at the client level, not at the AcquireToken (request) level?

**Comment by r-delgadillo** [comment]
> Does this always repro? Or do you have to have a few tokens in the cache already?
> 
> Possible workaround: Can you try to specify the authority only at the client level, not at the AcquireToken (request) level?

@bgavrilMS we are seeing this error pretty frequently - this graph shows the number of occurences in our PPE environment over the last day:
![image](https://user-images.githubusercontent.com/35089121/186950328-0ac71e0f-9fdb-4e34-8a10-08115dd9c402.png)

We need to be able to pass the `authority` at the request level because the client's authority is set to "https://login.microsoftonline.comorganizations" but when we receive a call we need to get the access/id/refresh tokens for a specific tenant (e.g. "https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47").

**Comment by jasonnutter** [comment]
@sameerag This seems similar to #4486, and there is comment indicating that particular issue wasn't completely solved: https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4486#issuecomment-1166832179

**Comment by r-delgadillo** [comment]
After turning off a feature flag that removes the flows for acquireTokenByRefreshToken & acquireTokenOnBehalfOf and performing a service restart, our acquireTokenByClientCredential method is working as expected again. Please let me know if you need anything else from our side. Thanks

**Comment by peterzenz** [comment]
@sameerag can you take a look at this and the similar issue that Jason has referenced above.

**Comment by bgavrilMS** [comment]
Just a hunch on this. The internal cache key under which we store access tokens is: 

`home_account_id-env-AccessToken-client_id-scoppes`

However, if client_credentials flow, there is no home_account_id, so the key looks like: 

`-env-AccessToken-client_id-scopes`

I wonder if there is a bug where `acquireTokenByClientCredential` cache lookup logic hits tokens from OBO ?

@r-delgadillo - what are the scopes that you use for OBO / auth_flow and what is the scope for client_credentials?

**Comment by r-delgadillo** [comment]
@bgavrilMS  For OBO flows, we are currently only getting an access token for MS graph (https://graph.windows.net/.default). 

For, `acquireTokenByRefreshToken` we are issuing the following:
- ["https://graph.microsoft.com/.default"]
- ["https://graph.windows.net/.default"]
- ["https://management.azure.com/.default"]

For, `acquireTokenByClientCredential` we are issuing the following tokens:
- ["https://management.azure.com/.default"]
- ["https://azure-devices-provisioning.net/.default"]
- ["https://iothubs.azure.net/.default"]


QQ. Would instantiating a new client every time mitigate the issue while we wait for a fix?

**Comment by bgavrilMS** [comment]
Yes @r-delgadillo, I think you can do that. You could:

- have 1 application object for `acquireTokenByRefreshToken ` and 1 for `acquireTokenByClientCredential`

If explicitly serialize (persist) your cache, e.g. you store the cache data to a file or smth, then just make sure the 2 application objects point to different files. Let me know if this is the case and I can advise further.

**Comment by bgavrilMS** [comment]
This would be a good bug to pick up for @Robbie-Microsoft, maybe @sameerag can act as advisor on this one?

**Comment by bgavrilMS** [comment]
TTD test: 

1. acquire token from ESTS `client_credentials` to ARM scope, store cache to a file
2. acquire token from ESTS `refresh_token` to ARM scope
3. repeat 1. It might repro the bug.

**Comment by r-delgadillo** [comment]
@bgavrilMS do you know who will be taking point on this and the potential timeline to get a fix out? Thank you

**Comment by sameerag** [comment]
@Robbie-Microsoft will be. We can reach out soon regarding the timeline, need to re-assess some of the other issues which are in the pipeline. Please watch out this space.

**Comment by r-delgadillo** [comment]
Adding comment to keep this item open.

**Comment by Robbie-Microsoft** [comment]
Apologies @r-delgadillo. I have begun working on this today.

**Comment by bgavrilMS** [comment]
The fix for this will be in the next release, 1.14.3 not 1.14.2

**Comment by cheenamalhotra** [comment]
It occurred to me today fyi.. 
```
[Error]: [ClientAuthError] Failed to find account: 
ClientAuthError: multiple_matching_tokens: The cache contains multiple tokens satisfying the requirements.
Call AcquireToken again providing more requirements such as authority or account.. - []
[Error]: MSAL: getToken call failed: [object Object] - []
```

The only change in my execution was I ran application as different user (Windows local user account), and added account there.. 

The API that caused error is: [TokenCache.getAccountByHomeId(accountId)](https://azuread.github.io/microsoft-authentication-library-for-js/ref/classes/_azure_msal_node.tokencache.html#getaccountbyhomeid) (when using PublicClientApplication) where `accountId` is in format: `00000000-0000-0000-77ed-ec3c7f46d602.9188040d-6c67-4c5b-b112-36a304b66dad`
